benchmarks/synchrobench/run.sh

```diff
@@ -11,121 +9,245 @@ function ctrl_c() {
+
+# Automatically picks the correct synchrobench JAR file
+jar_file_path=$(find "$(pwd)" -name "oak-benchmarks-synchrobench-*.jar" | grep -v "javadoc" | xargs ls -1t |  head -1)
+# Pipes breakdown:
+#  1. find the synchrobench JAR file (show full path)
+#  2. exclude the JAR with the javadoc
+#  3. sort by modification date (most recent first)
+#  4. take the first result (the most recent JAR file)
+
+# Iterate on the cartesian product of these arguments (space separated)
+test_scenarios=${!scenarios[*]}
+test_benchmarks=${!benchmarks[*]}
+test_thread="01 04 08 12 16 20 24 28 32"
+test_size="10000000"
+test_gc="default"
+test_java_modes="server"
+test_writes="0"
+
+# Defines the key size
 keysize="100"
-valuesize="1000"
-#writes="0 50"
-writes="0"
-warmup="0"
-iterations="5"
-duration="30000"
-benchs="JavaSkipListMap OakMyBufferMap OffHeapList"
-#gcAlgorithms="-XX:+UseParallelOldGC -XX:+UseConcMarkSweepGC -XX:+UseG1GC"
-
 
-while getopts "o:d:i:w:s:b:" opt; do
-  case ${opt} in
-    o ) output=$OPTARG;;
-    d ) duration=$(($OPTARG * 1000));;
-    i ) iterations=$OPTARG;;
-    w ) warmup=$OPTARG;;
-    s ) size=$OPTARG;;
-    b ) benchs=$OPTARG;;
-    \? )
-      echo "Invalid Option: -$OPTARG" 1>&2
-      exit 1
-      ;;
-    : )
-      echo "Invalid Option: -$OPTARG requires an argument" 1>&2
-      exit 1
-      ;;
-  esac
-done
+# Defines the value size
+valuesize="1000"
 
-declare -A heap_limit=(["OakMyBufferMap"]="12g"
-                       ["OffHeapList"]="12g"
-                       ["JavaSkipListMap"]="36g"
-                      )
+# Defines the number of warm-up (not measured) iterations.
+warmup="0"
 
-declare -A direct_limit=(["OakMyBufferMap"]="24g"
-                         ["OffHeapList"]="24g"
-                         ["JavaSkipListMap"]="0g"
-                        )
+# Defines the number of measured iterations.
+iterations="5"
 
-if [ ! -d "${output}" ]; then
-  mkdir $output
-else
-  rm -rf ${output}/*
-fi
+# Defines the test runtime in milliseconds.
+duration="30000"
 
+# Defines the sampling range for queries and insertions.
+range_ratio="2"
 
-###############################
-# records all benchmark outputs
-###############################
-
-declare -A scenarios=(
-                      ["4a-put"]="-a 0 -u 100"
-                      ["4b-putIfAbsentComputeIfPresent"]="--buffer -u 0 -s 100 -c"
-                      ["4c-get-zc"]="--buffer"
-                      ["4c-get-copy"]=""
-                      ["4d-95Get5Put"]="--buffer -a 0 -u 5"
-                      ["4e-entrySet-ascend"]="--buffer -c"
-                      ["4e-entryStreamSet-ascend"]="--buffer -c --stream-iteration"
-                      ["4f-entrySet-descend"]="--buffer -c -a 100"
-                      ["4f-entryStreamSet-descend"]="--buffer -c -a 100 --stream-iteration"
-                     )
+# There is no need to set different direct limit for each test because the JVM does not pre-allocate off-heap space.
+direct_limit="40g"
```

Comment:
I don't see how the way it was before helped you test different memory managers. It only had a different setting per implementation.

Reply:
Note that in all current implementations, we do not handle out-of-memory exceptions.
So reducing this size in different implementations will only make some of the tests to fail, which is not the purpose of benchmarking.
IMO, this kind of verifications should be reserved for unit testing.

This is different from the main heap size since the JVM maximal heap size has an effect on the performance even without yielding out-of-memory exceptions.

Reply:
Note 2: the benchmark monitors the JVM's memory usage, so it is still possible to compare how different MM handles the memory.